working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+  - Any function type
+  - Any class type or generic class type
+  - `FutureOr<S>` where `S` is non-nullable
+  - `X extends S` where `S` is non-nullable
+  - `X & S` where `S` is non-nullable
+
+Note that there are types which are neither nullable nor non-nullable.  For
+example `X extends T` where `T` is nullable is neither nullable nor
+non-nullable.
+
+We say that a type `T` is **potentially nullable** if `T` is not non-nullable.
+Note that this is different from saying that `T` is nullable.  For example, a
+type variable `X extends Object?` is a type which is potentially nullable but
+not nullable.
+
+We say that a type `T` is **potentially non-nullable** if `T` is not nullable.
+Note that this is different from saying that `T` is non-nullable.  For example,
+a type variable `X extends Object?` is a type which is potentially non-nullable
+but not non-nullable.
+
+It is an error to call a method, setter, getter or operator on an expression
+whose type is potentially nullable and not `dynamic`, except for the methods,
+setters, getters, and operators on `Object`.
+
+It is an error to read a field or tear off a method from an expression whose
+type is potentially nullable and not `dynamic`, except for the methods and
+fields on `Object`.
+
+It is an error to call an expression whose type is potentially nullable and not
+`dynamic`.
+
+It is an error for an instance field with no static initializer to have a
+potentially nullable type unless the variable or field is marked with the `lazy`
+modifier.
+
+It is an error if a local variable that is potentially non-nullable and is not
+marked `lazy` is used before it is definitely assigned (see Definite Assignment
+below).
+
+It is an error if a method, function, getter, or closure with a potentially
+non-nullable return type does not definitely complete (see Definite Completion
+below).
+
+It is an error if an optional parameter (named or otherwise) with no default
+value has a potentially non-nullable type.
+
+It is an error if a required named parameter has a default value.
+
+It is an error if a named parameter that is part of a `required` group is not
+bound to an argument at a call site.
+
+It is an error to call the default `List` constructor with a length argument and
+a type argument which is potentially non-nullable.
+
+For the purposes of errors and warnings, the null aware operators `?.` and `?..`
+are checked as if the receiver of the operator had non-nullable type.
+
+It is an error for a class to extend, implement, or mixin a type of the form
+`T?` for any `T`.
+
+It is an error for a class to extend, implement, or mixin the type `Never`.
+
+It is an error to call a method, setter, or getter on a receiver of static type
+`Never` (including via a null aware operator).
+
+It is an error to apply an expression of `Never` in the function position in a
+function call.
+
+It is an error if the static type of `e` in the expression `throw e` is
+potentially nullable.
```

Comment:
We can drop this. There is no actual problem with throwing Null.
This doesn't even allow us to disallow catching potentially nullable types, or even nullable types, because we don't have a type-constructor that makes a type non-nullable, and we would want:
```dart
foo<T>(Object o) { 
  try { 
     if (o != null) throw o;
  } on T catch (e) {
     Expect.identical(o, e);
  }
}
```
to work, and there is no type we can use instead of the potentially nullable `T` there.

Reply:
> There is no actual problem with throwing Null.

Why do you say there's no problem throwing null?  Doing so causes a NullThrownError, which is the "throw" equivalent of a null-pointer error.  Is this really ever what the user wanted?

> This doesn't even allow us to disallow catching potentially nullable types, or even nullable types

Why can't we disallow catching nullable types?  There's no reason to allow `on int? catch (e) {` because `e` will never be `null`, no matter what.  I guess technically you could want to do `on FutureOr<int?> catch(e)`, but.... really?

Reply:
I mean that if we choose to remove the artificial restriction on throwing `null`, we could. Things would still work. We added the rule because some people didn't think it worked well with the catch syntax at the time, `catch (SomeError e)`, where `SomeError` looked like a type annotation and would therefore be assumed to catch null. To avoid confusion, you could no longer throw `null`. The current syntax `on SomeError catch (e)` doesn't have this problem so we could just allow throwing `null` again.

If we keep throwing `null` disallowed, then perhaps we can disallow catching (some) definitely nullable types, but not potentially nullable types. We probably must allow `on X`, a type variable, or `on dynamic`. I don't see any big advantage to disallowing `on Object?`, the `on` clause of `catch` blocks generally works like an `is` check, so we can allow anything there that an `is` check allows. I don't think adding restrictions is worth the complexity, and I doubt we'll catch any errors with those restrictions.

Reply:
> If we keep throwing `null` disallowed

Are we considering changing this?  I certainly wasn't.

Reply:
I was always considering it (I never cared for the restriction), but with NNBD it has begun to actually matter. We need to decide whether `try { ... } on int? { ... }` is allowed, whether it's useful, and how that affects the type of `e` in `try { ... } catch (e) { ... }`.

On one hand, disallowing `null` in some situations is exactly what NNBD is for, and if we think that throwing `null` is most likely an error, we should keep disallowing it. (That is probably the case, but I want to make sure).

On the other hand, we now have a way to distinguish between `on int` and `on int?`, so any compunctions we had about whether the `on` type should or shouldn't match `null` is now gone. If that was the only reason to not allow throwing `null` (it was the only one given at the time, but maybe just because it was sufficient), then we could just allow throwing `null`.

If we allow throwing `null`, then the default `on` clause of a `try` would be `on Object?`. That also means that `try { ... } catch (e, s) { ... }` would give `e` the static type `Object?`, but `Object?` is probably not really distinguishable from `Object` in practice (they allow the same methods, and only differ in assignability in that `Object?` is not assignable to `Object`).

If we disallow throwing `null`, then we also disallow potentially nullable typed expressions in throws (the context type of the `throw` operand is `Object`).

Do we also disallow nullable types in `try`/`on` clauses? That is, can you write:
```dart
try {
  ...
} on Exception? catch (e, s) {
   // e has static type Exception?
}
```
or do we disallow it because it's known to be unnecessary.
Would we also disallow potentially nullable types? (Likely yes, but it means that:
```dart
T catcher<T>(void Function() thrower) {
  try {
    thrower();
  } on T catch (e) {
    return e;
  }
  throw ArgumentError.value(thrower, "thrower", "Didn't throw");
}
```
would be illegal, and you have to make it `T catcher<T extends Object>(...)`. That's also more correct, so it's not necessarily a problem.


All in all, I think we should think about this just once more, just to be sure. 
It's probably correct that `throw null` is a mistake, a value accidentally being `null`, and disallowing it is doing the user a service.

Reply:
Ok, filed [an issue](https://github.com/dart-lang/language/issues/385) on this, let's discuss and resolve it there.